### ISSUE TITLE
auth: permissions cache: prevent stressing the particular shard in a 'single user + many tables' case. 

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#2589]
*Installation details*
Scylla version (or git commit hash): bf49dd8aa1eaf4ae3b5a8ade3104e25803b8eff2

*Description*
If there is a single "hot" user with many tables it would translate into a "hot" shard with a high reads rate coming from the permissions cache.

This is caused the fact that permissions cache is refreshing its values by reading the `system_auth.permissions` table. This table's schema is as follows:

```
CREATE TABLE system_auth.permissions (
    username text,
    resource text,
    permissions set<text>,
    PRIMARY KEY (username, resource)
) WITH CLUSTERING ORDER BY (resource ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = '{"keys":"ALL","rows_per_partition":"ALL"}'
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 7776000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

The partition key is a `username` while the permissions cache key is a `std::pair<authenticated_user, data_resource>`. This means that the cache will issue as many reads as the number of the actively accessed tables. However if `username` is going to be the same then all these queries are going to land on the same shard of a specific replica.

One way to fix this is to make a `resource` a part of a partition key.

**Comment by vladzcloudius** [comment]
A straight forward change (simply set the PK to ((username, resource)) )is not going to work since we need to query the CF data by 'resource' value (from the LIST `cqlsh` command flow).

So, one way to address this is to add an additional column 'resource_for_pk' that is going to have the same value as a 'resource' column and be a part of a partition key. This however is going to break the backwards compatibility and the same UPDATE/INSERT query will no longer work on the original `system_auth.permissions`.

@slivne proposed the elegant solution for that:
   - Introduce a new CF (`permissions1`) in the `system_auth` KS with the new schema as described above.
   - Keep the `system_auth.permissions` as is.
   - From now on add new permissions into the new table.
   - When querying for permissions, first, query the new CF and if not found, query the old one.

**Comment by slivne** [comment]
My suggestion was ro add a materialized view like solution

Add another table with the rwquested modified partition key

When inserting data insert to both when deleting remive from both

That will allow
- cqlsh to stay as is (and any other tool that queries directly)
- internal auth queries to be written to utilize the modified table with
the modified partition key

(Once we have mv complete we can remove the redundant code)

**Comment by vladzcloudius** [comment]
Sorry, l got you wrong then... ;) I updated the confusing comment above.

**Comment by vladzcloudius** [comment]
@slivne writing into two tables seems a little risky - what if the first write succeeds and the second doesn't?

**Comment by vladzcloudius** [comment]
@slivne BTW, I think cqlsh doesn't run explicit "SELECT" on `system_auth.permissions` CF but rather uses a "LIST PERMISSIONS" cql command, which invokes the appropriate "SELECT", so my proposal should not influence `cqlsh` as well.

**Comment by vladzcloudius** [comment]
@slivne I have to agree that MVs would be the best option here. For now we can have a special CF for the permissions cache with the composite partition key = `((user, resource))` that would be written only when we load a particular user + resource value and it's not present in this new table. The code that updates/deletes entries from the `system_auth.permissions` would simply truncate the new CF to keep things simple. We know that updates in system_auth.permissions are rare so the negative impact should be very low.

**Comment by vladzcloudius** [comment]
@avikivity @hakuch 

As discussed on the mailing list, currently the most promising direction seems to be adding an RPC that would drop the whole permissions cache content on all shards on the destination Node and invoking it  on all cluster Nodes every time the permissions are changed. 

This also means that the `system_auth` has to have RF=ALL and the update command has to be performed with CL=ALL. Otherwise the RPC won't be able to ensure that the new permissions are going to be loaded after it completes.

The networking disconnects should not be tolerated - if do not know for sure that the Node is down and we can't complete the RPC we should fail the whole permissions update command. @asias Is there a way do differentiate between the Node that is stopped and the one that is temporarily unavailable. I think they would "look" the same. If that's the case we should demand all cluster Nodes to be up and available when we update the permissions which doesn't seem too much to demand since we are talking about the security and these changes should be very rare.

After the above is implemented we may disable the "refresh" part of the permissions cache completely.

**Comment by kostja** [comment]
@nuivall yet another request for the same solution - full cache. Please note that the more duplicates a problem has the more pressing it is in the field.
